test/functional/not_fn.cpp

```diff
@@ -0,0 +1,353 @@
+/*
+  SLB.Functional
+
+  Copyright Michael Park, 2017
+  Copyright Agustin Berge, 2017
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#include <slb/functional.hpp>
+
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+#define CATCH_CONFIG_MAIN
+#include "../catch.hpp"
+
+#define CHECK_NESTED(...)                                                      \
+  do {                                                                         \
+    INFO(__FILE__ "(" << __LINE__ << "): " #__VA_ARGS__);                      \
+    check_##__VA_ARGS__;                                                       \
+  } while (false)
+
+// [func.not_fn], function template not_fn
+
+// template <class F> unspecified not_fn(F&& f);
+
+struct Boolean {
+  int ret;
+  Boolean(int val) : ret(val) {}
+
+  int operator!() const noexcept { return ret; }
+};
+
+struct BooleanRef {
```

Comment:
Why is this called `BooleanRef`...?

Reply:
It used to hold a reference to an `int`, member of `C`.

Is `BooleanObj` any better?